server/src/rate_limit/mod.rs

```diff
@@ -2,17 +2,182 @@ pub mod rate_limiter;
+
+  pub fn post(&self) -> RateLimited {
+    self.kind(RateLimitType::Post)
+  }
+
+  pub fn register(&self) -> RateLimited {
+    self.kind(RateLimitType::Register)
+  }
+
+  fn kind(&self, type_: RateLimitType) -> RateLimited {
+    RateLimited(self.0.clone(), type_)
+  }
+}
+
+impl RateLimited {
+  pub async fn wrap<T, E>(
+    self,
+    ip_addr: String,
+    fut: impl Future<Output = Result<T, E>>,
+  ) -> Result<T, E>
+  where
+    E: From<failure::Error>,
+  {
+    let rate_limit: RateLimitConfig = actix_web::web::block(move || {
+      // needs to be in a web::block because the RwLock in settings is from stdlib
+      Ok(Settings::get().rate_limit) as Result<_, failure::Error>
+    })
+    .await
+    .map_err(|e| match e {
+      actix_web::error::BlockingError::Error(e) => e,
+      _ => APIError::err("Operation canceled").into(),
+    })?;
+
+    // before
+    {
+      let mut limiter = self.0.lock().await;
+
+      match self.1 {
+        RateLimitType::Message => {
+          limiter.check_rate_limit_full(
+            self.1,
+            &ip_addr,
+            rate_limit.message,
+            rate_limit.message_per_second,
+            false,
+          )?;
+
+          return fut.await;
+        }
+        RateLimitType::Post => {
+          limiter.check_rate_limit_full(
+            self.1.clone(),
+            &ip_addr,
+            rate_limit.post,
+            rate_limit.post_per_second,
+            true,
+          )?;
+        }
+        RateLimitType::Register => {
+          limiter.check_rate_limit_full(
+            self.1,
+            &ip_addr,
+            rate_limit.register,
+            rate_limit.register_per_second,
+            true,
+          )?;
+        }
+      };
+    }
+
+    let res = fut.await;
+
+    // after
+    {
+      let mut limiter = self.0.lock().await;
+      if res.is_ok() {
+        match self.1 {
+          RateLimitType::Post => {
+            limiter.check_rate_limit_full(
+              self.1,
+              &ip_addr,
+              rate_limit.post,
+              rate_limit.post_per_second,
+              false,
+            )?;
+          }
+          RateLimitType::Register => {
+            limiter.check_rate_limit_full(
+              self.1,
+              &ip_addr,
+              rate_limit.register,
+              rate_limit.register_per_second,
+              false,
+            )?;
+          }
+          _ => (),
+        };
+      }
+    }
+
+    res
+  }
+}
+
+impl<S> Transform<S> for RateLimited
+where
+  S: Service<Request = ServiceRequest, Response = ServiceResponse, Error = actix_web::Error>,
+  S::Future: 'static,
+{
+  type Request = S::Request;
+  type Response = S::Response;
+  type Error = actix_web::Error;
+  type InitError = ();
+  type Transform = RateLimitedMiddleware<S>;
+  type Future = Ready<Result<Self::Transform, Self::InitError>>;
+
+  fn new_transform(&self, service: S) -> Self::Future {
+    ok(RateLimitedMiddleware(self.clone(), service))
+  }
+}
```

Comment:
I have no idea what this is doing, other than actix needs for middleware. Where are the docs for this so I can understand how you made this?

Its kinda frustrating because rust-analyzer can't even trace instances of these functions: wrap, poll_ready, call, new_transform... I'm okay w/ it obviously if its documented somewhere, but it makes debugging issues, or adding to this later a little scary.

Reply:
The way I assume this works internally is this
- You create the Transform type, and hand that to Actix Web's `.wrap` method.
- When a request is made, Actix Web calls `.new_transform` on your transform type to build the middleware stack. the `service: S` here is the middleware mounted inside this one. That middleware could be the actix-provided route handler, or another one you've added to the stack.

The docs for building middlewares can be found here: 
- Transform: https://docs.rs/actix-web/2.0.0/actix_web/dev/trait.Transform.html
- Service: https://docs.rs/actix-web/2.0.0/actix_web/dev/trait.Service.html